jctools-experimental/src/main/java/org/jctools/uc/FlatCombining.java

```diff
@@ -0,0 +1,181 @@
+package org.jctools.uc;
+
+import org.jctools.stacks.WFIStack;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.LockSupport;
+
+/**
+ * FlatCombining is a <em>universal construct</em> which can provide thread-safe and concurrent access to an
+ * <em>arbitrary</em> sequential data structure.
+ * <p>
+ * The only requirement is that all accesses to the data structure must be delivered as
+ * {@linkplain Operation operations} through the {@link #apply(Operation)} method.
+ * <p>
+ * Each operation will only execute exactly once, and there will only be a single copy of the underlying data structure.
+ * <p>
+ * The only consequcne is that the {@link #apply(Operation)} method may block.
+ * <p>
+ * This is conceptually similar to wrapping the data structure in a mutex, but the {@code FlatCombining} implementation
+ * will scale positively with the level of concurrency, while contended locks typically scales negatively.
+ * Negative scalability is when the system-wide throughput drops as the number of active threads increases.
+ * <p>
+ * The Flat Combining concept is described by the paper
+ * <a href="https://www.cs.bgu.ac.il/~hendlerd/papers/flat-combining.pdf">
+ * Flat Combining and the Synchronization-Parallelism Tradeoff</a>
+ *
+ * @param <T> The type of data managed by this flat combining instance.
+ */
+public class FlatCombining<T> {
+    private final WFIStack<OpNode> ops;
+    private final AtomicBoolean lock;
+    private final T data;
+
+    public FlatCombining(T sequentialData) {
+        ops = new WFIStack<OpNode>();
+        lock = new AtomicBoolean();
+        data = sequentialData;
+    }
+
+    /**
+     * Apply the given operation on the underlying data structure. This call may block. Multiple threads can calls
+     * this at the same time. The operations will be applied with linearizable consistency.
+     * <p>
+     * The operation will only be applied once to the data structure, but there is no guarantee that it will be
+     * applied by the calling thread.
+     *
+     * @param operation The operation to apply to the underlying data structure.
+     * @param <R>       The type of return value produced by the operation, if any.
+     * @param <E>       The type of exception the operation might throw, if any.
+     * @return The result returned by the operation.
+     * @throws E                    If the operation throws an exception.
+     * @throws InterruptedException If the apply method ends up blocking but the thread is interrupted.
+     */
+    public <R, E extends Exception> R apply(Operation<T, R, E> operation) throws E, InterruptedException {
+        OpNode op = enqueue(operation);
+
+        for (; ; ) {
+            if (tryLock()) {
+                Iterable<OpNode> nodes = combineAndApplyAllOperations();
```

Comment:
Although counter-intuitive I think that would be nice to have a version of `combineAndApplyAllOperations` that allow to limit the amount of work to be performed (the simpler way in the number of combined operations, but can think to other metrics, or by using a cooperative suspention of some form eg `ControlledRunnable`?) to allow other(s) Threads to pick remaining jobs (if the combiner thread isn't alone).

Reply:
Thinking it twice probably the limit in the combined list of pending ops work only if the list is FIFO. Need to think about it better